### ISSUE TITLE
OAuth 1, wrong OAuth redirect

### ISSUE THREAD

**Issue by rmonnier9** [simov/grant#249]
Hi !

I'm trying to use grant with Sellsy (https://go.sellsy.com/fr/).

Here is my configuration:

```js
{
    sellsy: {
      enabled: true,
      request_url: "https://apifeed.sellsy.com/0/request_token",
      authorize_url: "https://apifeed.sellsy.com/0/oauth/access_token",
      access_url: "https://apifeed.sellsy.com/0/login.php?",
      oauth: 1,
      key: "XXX",
      secret: "XXX",
      callback: "http://localhost:1337/auth/sellsy/callback",
    }
}
```

The problem is that the url where grant redirects the user is:
http://localhost:1337/connect/authentification_url=https://apifeed.sellsy.com/0/login.php&oauth_token=0a1bfe71494c3fe920c2adfd9bb5853dc2d1f&oauth_token_secret=395dsqfqfdff8daae8de6da5879ed053&oauth_callback_confirmed=true

It doesn't work but if I remove the first part ('http://localhost:1337/connect/authentification_url=') and if I replace the '&' after login.php with an '?' the redirection works.

What am I missing with the configuration?

Thanks a lot

**Comment by simov** [comment]
I think it should be this instead:

```json
{
  "sellsy": {
    "request_url": "https://apifeed.sellsy.com/0/request_token",
    "authorize_url": "https://apifeed.sellsy.com/0/login.php",
    "access_url": "https://apifeed.sellsy.com/0/oauth/access_token",
    "oauth": 1,
    "key": "XXX",
    "secret": "XXX",
    "callback": "http://localhost:1337/auth/sellsy/callback"
  }
}
```

You setup the OAuth URLs incorrectly. However, that may still not work because Sellsy expects `PLAINTEXT` as `signature_method` and that is currently not exposed in Grant.

So in case things doesn't work even with the correct URLs, can you go to your `node_modules/grant` folder and update these two places manually:

- https://github.com/simov/grant/blob/master/lib/flow/oauth1.js#L35-L36
- https://github.com/simov/grant/blob/master/lib/flow/oauth1.js#L108-L109

It should be `if (provider.sellsy)`. If that works then I'm going to add it in Grant as well.

**Comment by rmonnier9** [comment]
Thanks for your answer !

I changed the URL order and add sellsy in the grant files.
I think it was necessary, but i'm still facing the first problem.

After using GET /connect/sellsy i've got a redirection to /connect/authentification_url=https://apifeed.sellsy.com/0/login.php&oauth_token=XXX&oauth_token_secret=395dsqfqfdff8daae8de6da5879ed053&oauth_callback_confirmed=true

I'm not able to understand where does location comes from at this line:
https://github.com/simov/grant/blob/master/lib/handler/koa-1.js#L42
This is where I see the wrong url for the first time.

**Comment by simov** [comment]
Can you confirm that:

1. You have set the `redirect_uri` of your OAuth app in Sellsy to `http://localhost:1337/auth/sellsy/callback`
2. You are using your browser to navigate to `/connect/sellsy` on your server
3. Your browser is redirecting you to `https://apifeed.sellsy.com/0/login.php` where you can login and authorize your app
4. You are being redirected back to your `callback` URL which is different than your `redirect_uri` -> `/auth/sellsy/callback`

**Comment by rmonnier9** [comment]
1. It seems I'm not able to configure the redirect_uri with sellsy or I have already done it (the next steps work anyway)
2. I use /connect/sellsy
3. Then my browser redirects me to ```http://localhost:1337/connect/authentification_url=https://apifeed.sellsy.com/0/login.php&oauth_token=XXX&oauth_token_secret=XXX&oauth_callback_confirmed=true``` (see screenshot of my API).

![image](https://user-images.githubusercontent.com/18701290/122650702-b3cd1a80-d134-11eb-9cf9-05e504fb0bae.png)


At this steps, if I manually remove the beginning ```http://localhost:1337/connect/authentification_url=``` and I add a question mark to```login.php&oauth_token``` (in order to get ```login.php?&oauth_token```) then I'm on Sellsy and I can login
4. I'm finally redirected to my callback url ```/integrations/sellsy/callback``` but i get this: ```http://localhost:3000/integrations/sellsy/callback?error=oauth_problem=signature_invalid&debug_sbs=POST&https%3A%2F%2Fapifeed.sellsy.com%2F0%2Foauth%2Faccess_token&oauth_consumer_key%3D9306cvcbdb9ce65b2ad734bc13c1e126137cf3f27%26oauth_nonce%3D902a9dcbb86ef49698f6deceeba1d50ec4%26oauth_signature_method%3DPLAINTEXT%26oauth_timestamp%3D1624123865%26oauth_token%3D4d408116fdbae02dce880729efd08861b8c0b645%26oauth_verifier%3Dfbcf5c4766e8656b5fj6bd4d092c35db3edecc1ae%26oauth_version%3D1.0```

**Comment by simov** [comment]
TBH I've no idea what that redirect is. In case this provider isn't officially supported yet, it's always a good idea to verify that Grant works with it first. You can use this [example](https://github.com/simov/grant/blob/master/examples/transport-session/koa.js). If that works then the question is why your Framework isn't working. Can you try that first.

**Comment by rmonnier9** [comment]
I've just tried with the example file and i've got the same, redirected to ```http://localhost:3000/connect/authentification_url=https://apifeed.sellsy.com/0/login.php&oauth_token=98e6f5fe8b5808e998ce172446a61b1bcbcf0116&oauth_token_secret=8290c91ad7e8ae2e11e1f7350e9759b993b30a86&oauth_callback_confirmed=true```

**Comment by rmonnier9** [comment]
I will try to use debugger and breakpoint to understand what's happening under the hood.

**Comment by rmonnier9** [comment]
It seems that the body of the answer to ```POST 'https://apifeed.sellsy.com/0/request_token'``` is ```'authentification_url=https://apifeed.sellsy.com/0/login.php&oauth_token=a9b037c8ff8bc830b2acc5eafa05d6dfe15cdc06&oauth_token_secret=4e1b880ac2f32812949b6c443dce4b9b86bb23e1&oauth_callback_confirmed=true'```

**Comment by simov** [comment]
I've no idea what that is TBH, this endpoint is expected to return `application/x-www-form-urlencoded` encoded body.

**Comment by rmonnier9** [comment]
At this step, Grant is expecting a json but sellsy returns a string:

![image](https://user-images.githubusercontent.com/18701290/122651576-1248c780-d13a-11eb-94ce-8d33d2d451f6.png)

So I think sellsy is not compatible.

**Comment by rmonnier9** [comment]
Thanks for your help anyway :)

**Comment by simov** [comment]
If I had access to that provider I could have make it work probably. Otherwise it's really hard to guess what the intent of the developers was. Does they have a free plan that I can use for testing?

**Comment by rmonnier9** [comment]
Yes they have.
- The documentation: https://api.sellsy.com/documentation/oauth
- The page to create an app: https://www.sellsy.fr/developer/my-apps

**Comment by rmonnier9** [comment]
I've tried to use sellsy with oauth-node and it works.

If you check at this file:
https://github.com/ciaranj/node-oauth/blob/master/lib/oauth.js#L545

They use query-string to extract the params from the url.

**Comment by simov** [comment]
> They have a free plan and you can ask to extend if you explain them that you are building an integration.

Oh I know how that usually goes trust me ..

> I've tried to use sellsy with oauth-node and it works.

That's interesting, might be some edge case that I don't know about.

**Comment by rmonnier9** [comment]
I've created a pull request that resolves the first problem with the wrong redirection:
https://github.com/simov/grant/pull/250
(it is a quick fix, we sure could find something more elegant)

There is now a second issue: the signature method is HMAC-SHA1.
In node-oauth, here is how they do it:
https://github.com/ciaranj/node-oauth/blob/master/lib/oauth.js#L214

**Comment by rmonnier9** [comment]
Then I understand why it still doesn't work:
it is because the session in koa has kept the unparsed ```output``` with querystring.

If you know where the session is saved, you could apply querystring before where i do it in my PR.

**Comment by rmonnier9** [comment]
This PR solves the issue:
https://github.com/simov/grant/pull/251

**Comment by simov** [comment]
Resolved in #251
